### PR TITLE
Initial Mid-level `FlightClient`

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Towards https://github.com/apache/arrow-rs/issues/3371

# Rationale for this change
I am building a FlightSQL client. There are many lower level details such as RecordBatch reconstruction that are not FlightSQL specific. Not only are they easier to test as a separate layer, the mid level client is useful on its own (for writing clients for other Arrow Flight based APIs) so I am contributing it here. 

More details are on  https://github.com/apache/arrow-rs/issues/3371

If this basic approach is accepted, I plan:
1. PRs with tests for do_get (this PR is already quite large)
2. PR to update the FlightSQL client to use `FlightClient`
3. File follow on tickets for the remaining APIS (e.g. do_action). I will likely not have time to implement them

# What changes are included in this PR?
1. Initial `FlightClient`, `FlightError` modules, originally written for IOx in https://github.com/influxdata/influxdb_iox/pull/6427
4. Documentation 
5. Integration test

# Are there any user-facing changes?
These are new features

cc @tustvold @Dandandan @avantgardnerio

### CODE REVIEW

File: arrow-flight/src/client.rs [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+    }
+}
+
+/// A stream of [`RecordBatch`]es from from an Arrow Flight server.
+///
+/// To access the lower level Flight messages directly, consider
+/// calling [`Self::into_inner`] and using the [`FlightDataStream`]
+/// directly.
+#[derive(Debug)]
+pub struct FlightRecordBatchStream {
```
Comment (alamb) [inline arrow-flight/src/client.rs:295]: This code is not tested yet. I plan to add tests for it as a follow on PR

**Comment by alamb** [comment]
Marking as draft as I work out conflicts with https://github.com/apache/arrow-rs/pull/3359

File: arrow-flight/tests/common/server.rs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        Default::default()
+    }
+}
+
+/// Implement the FlightService trait
+#[tonic::async_trait]
+impl FlightService for TestFlightServer {
+    type HandshakeStream = Pin<
+        Box<dyn Stream<Item = Result<HandshakeResponse, Status>> + Send + Sync + 'static>,
+    >;
```
Comment (tustvold) [inline arrow-flight/tests/common/server.rs:129]:
```suggestion
    type HandshakeStream = BoxStream<'static, Result<HandshakeResponse, Status>>;
```
And the same for the ones below

File: arrow-flight/src/client.rs [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+    /// # }
+    /// ```
+    pub async fn do_get(&mut self, ticket: Ticket) -> Result<FlightRecordBatchStream> {
+        let request = self.make_request(ticket);
+
+        let response = self
+            .inner
+            .do_get(request)
+            .await
+            .map_err(FlightError::Tonic)?
```
Comment (tustvold) [inline arrow-flight/src/client.rs:211]:
```suggestion
            .await?
```

File: arrow-flight/src/client.rs [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+        };
+
+        // apply headers, etc
+        let request = self.make_request(stream::iter(vec![request]));
+
+        let mut response_stream = self
+            .inner
+            .handshake(request)
+            .await
+            .map_err(FlightError::Tonic)?
```
Comment (tustvold) [inline arrow-flight/src/client.rs:150]:
```suggestion
            .await?
```

File: arrow-flight/src/client.rs [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+        &mut self,
+        descriptor: FlightDescriptor,
+    ) -> Result<FlightInfo> {
+        let request = self.make_request(descriptor);
+
+        let response = self
+            .inner
+            .get_flight_info(request)
+            .await
+            .map_err(FlightError::Tonic)?
```
Comment (tustvold) [inline arrow-flight/src/client.rs:267]:
```suggestion
            .await?
```

File: arrow-flight/src/client.rs [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+
+        let mut response_stream = self
+            .inner
+            .handshake(request)
+            .await
+            .map_err(FlightError::Tonic)?
+            .into_inner();
+
+        if let Some(response) = response_stream.next().await {
+            let response = response.map_err(FlightError::Tonic)?;
```
Comment (tustvold) [inline arrow-flight/src/client.rs:154]:
```suggestion
        if let Some(response) = response_stream.next().await.transpose()? {
```

File: arrow-flight/src/client.rs [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+    ///
+    /// See [`FlightClient`] docs for an example.
+    pub async fn handshake(&mut self, payload: impl Into<Bytes>) -> Result<Bytes> {
+        let request = HandshakeRequest {
+            protocol_version: 0,
+            payload: payload.into(),
+        };
+
+        // apply headers, etc
+        let request = self.make_request(stream::iter(vec![request]));
```
Comment (tustvold) [inline arrow-flight/src/client.rs:144]:
```suggestion
        let request = self.make_request(stream::once(ready(request)));
```

**Review by tustvold — COMMENTED** [review]
Looks good to me

File: arrow-flight/tests/common/server.rs (hunk 2 above)
Comment (alamb) [reply arrow-flight/tests/common/server.rs:129]: 👍  -- this is actually copy/paste from the example server

Comment (alamb) [reply arrow-flight/tests/common/server.rs]: in 901bd7556

**Review by tustvold — APPROVED** [review]

**Pull request merged**